python/revng/internal/api/synchronizer.py

```diff
@@ -0,0 +1,300 @@
+            r"(?P<region>[^+]*)\+(?P<host>[^/]*)/(?P<bucket>[^/]+)(?P<path>/.*)?$",
+            url,
+        )
+        if match_obj is None:
+            raise ValueError("S3 endpoint invalid")
+
+        self.botocore_session = BotocoreSession()
+        self.botocore_session.set_credentials(match_obj["username"], match_obj["password"])
+        config = botocore.config.Config(s3={"addressing_style": "path"})
+        session = boto3.session.Session(botocore_session=self.botocore_session)
+        proto = "https" if match_obj["proto"] == "s3s" else "http"
+        self.client = session.client(
+            "s3",
+            endpoint_url=f"{proto}://{match_obj['host']}",
+            region_name=match_obj["region"],
+            config=config,
+        )
+        self.bucket = match_obj["bucket"]
+        if match_obj["path"] is not None:
+            self.path = match_obj["path"][1:]
+        else:
+            self.path = ""
+
+        self.last_save_event: Event | None = None
+        self.last_save_event_lock = Lock()
+        self.last_file_list: dict[str, FileData] = {}
+        self.last_file_list_lock = Lock()
+
+    def load(self, path: Path) -> bool:
+        # Check that the provided path is not a file
+        assert not path.exists() or path.is_dir()
+        path.mkdir(parents=True, exist_ok=True)
+
+        # Download index.yml from S3 and parse it
+        index = self._get_index()
+        # If the index is empty do nothing
+        if len(index) == 0:
+            return True
+
+        # Generate the `last_file_list`, with the data retrieved from index.yml
+        self.last_file_list = {k: FileData(v) for k, v in index.items()}
+        # Download all the files mentioned in index.yml in the specified
+        # resume directory
+        queue = [
+            executor.submit(self._download_file, value, path / key) for key, value in index.items()
+        ]
+        # Actually wait for all the downloads to finish
+        return gather_bools(queue, "Failed to load")
+
+    def save(self, path: Path) -> bool:
+        # Check that the provided path is not a file
+        assert not path.exists() or path.is_dir()
+        # Early exit, if the directory does not exist
+        if not path.exists():
+            return True
+
+        # Replace the supplied directory, using a temporary directory in the
+        # same parent directory. The temporary directory is created in the
+        # parent directory to mitigate crossing filesystem boundaries as much
+        # as possible (otherwise the `os.replace` call below will fail).
+        temp_dir = TemporaryDirectory(dir=path.parent)
+        temp_dir_path = Path(temp_dir.name)
+        # Moves the directory to the temporary directory, atomically (if the
+        # filesystem is POSIX compliant). Will fail if it cannot do so.
+        os.replace(path, temp_dir.name)
+        # Remake the provided directory
+        os.mkdir(path)
```

Comment:
If at the beginning we had:

```
parent/resume-dir
```

After this line we will have

```
parent/temporary-dir/resume-dir
parent/resume-dir
```

Is this correct? If so, put a comment similar to the above.

Reply:
No, after that line we have:
```
parent/temporary-dir
parent/resume-dir
```

So, for instance, `context/model.yml` goes from `parent/resume-dir/context/model.yml` to `parent/temporary-dir/context/model.yml`

Reply:
So `os.replace` allows moving over an existing directory?
I guess it fails if the target directory is not empty?

Reply:
> So os.replace allows moving over an existing directory?

Yes

> I guess it fails if the target directory is not empty?

Exactly

Reply:
Added some more comments explaining what happens